src/Chapter3.hs

```diff
@@ -1006,6 +1128,92 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KnightAction
+    = KnightAttack
+    | KnightCastSpell Int
+    | KnightDrinkPotion Int
+    deriving Show
+
+data MonsterAction
+    = MonsterAttack
+    | MonsterRunAway
+    deriving Show
+
+data Knight = Knight
+    { knightAttack :: Int
+    , knightDefense :: Int
+    , knightHealth :: Int
+    , knightActions :: [KnightAction]
+    } deriving Show
+
+data Monster = Monster
+    { monsterAttack :: Int
+    , monsterHealth :: Int
+    , monsterActions :: [MonsterAction]
+    } deriving Show
+
+class (Show a) => Fighter a where
+    duel :: (Fighter b) => a -> b -> (a, b)
+    sufferDamage :: a -> Int -> a
+    isDead :: a -> Bool
+
+instance Fighter Knight where
+    isDead knight = knightHealth knight <= 0
+
+    sufferDamage knight points = knight
+        { knightDefense = knightDefense knight - defenseDamage
+        , knightHealth = knightHealth knight - healthDamage
+        }
+      where
+        defenseDamage = round $ 0.8 * fromIntegral points
+        healthDamage = round $ 0.2 * fromIntegral points
+
+    duel knight opponent
+        | null actions = (knight, opponent)
```

Comment:
Instead of checking `actions` with `null` and using partial functions `head` and `tail` in `where`, it's better to pattern match directly on the list:

```haskell
duel knight opponent = case knightActions knight of
    [] -> (knight , opponent)
    curAction:actions -> case curAction of
        ... -> ...
```

Reply:
:+1: Thanks, I learned something new! Woohoo! \o/